akka-http-core/src/main/scala/akka/http/impl/model/parser/UriParser.scala

```diff
@@ -184,12 +188,16 @@ private[http] final class UriParser(val input: ParserInput,
 
   // http://www.w3.org/TR/html401/interact/forms.html#h-17.13.4.1
   def query: Rule1[Query] = {
-    def part = rule(
-      clearSBForDecoding() ~ oneOrMore('+' ~ appendSB(' ') | `query-char` ~ appendSB() | `pct-encoded`) ~ push(getDecodedString())
+    def keyPart = rule(clearSBForDecoding() ~
+      oneOrMore('+' ~ appendSB(' ') | `query-key-char` ~ appendSB() | `pct-encoded`) ~ push(getDecodedString())
+        | push(""))
+
+    def valuePart = rule(clearSBForDecoding() ~
```

Comment:
Can be simplified to

```scala
    def part(`query-char`: CharPredicate) =
      rule(clearSBForDecoding() ~
        oneOrMore('+' ~ appendSB(' ') | `query-char` ~ appendSB() | `pct-encoded`) ~ push(getDecodedString())
        | push(""))

    def keyValuePair: Rule2[String, String] = rule {
      part(`query-key-char`) ~ ('=' ~ part(`query-value-char`) | push(Query.EmptyValue))
    }
```

(then the error message expectations need to be adapted as well again).